### PR TITLE
WIP: Creates a request log

### PR DESCRIPTION
**Pull request by sradack**
## Changes proposed in this PR

Create a request log.  In scope:

- Log proxied requests
- One request per line
- Structured log format suitable for indexing

## Why are we making these changes?

A request log is a more compact, structured form of logging than what we do currently.  

## TODO
- Plug into logging framework
- Handle WS requests
- Decide on structured output format
- Tests

## Sample log
`{:termination-state :success, :path "/sleep", :service-id "waiter-service-exampleservice-73e3607e7980ff4b57cb2c455cd0af89", :service-name "example-service", :backend-latency 83, :instance-port 10005, :method "GET", :cid "262a9d71848d2d-37f288159dba2c2a", :total-latency 6427, :bytes-streamed 15, :discovery-latency 49, :instance-id "262a9dbad2d076-16de7b48e62c412c", :overhead-latency 6344, :instance-latency 6343, :host "example.com", :status 400, :instance-proto "http", :principal "sradack", :query-string "sleep-ms=50", :timestamp "2018-02-22T09:03:50.929-06:00", :scheme :http, :instance-host "127.0.0.5", :service-version "2", :metric-group "example-metric-group"}`

`{:termination-state :success, :path "/sleep", :service-id "waiter-service-exampleservice-73e3607e7980ff4b57cb2c455cd0af89", :service-name "example-service", :backend-latency 56, :instance-port 10005, :method "GET", :cid "262aa0383a53e9-56136a954ba05930", :total-latency 59, :bytes-streamed 15, :discovery-latency 1, :instance-id "262a9dbad2d076-16de7b48e62c412c", :overhead-latency 3, :instance-latency 2, :host "example.com", :status 400, :instance-proto "http", :principal "sradack", :query-string "sleep-ms=50", :timestamp "2018-02-22T09:04:02.852-06:00", :scheme :http, :instance-host "127.0.0.5", :service-version "2", :metric-group "example-metric-group"}`

`{:termination-state :backend-error, :path "/die", :service-id "waiter-service-exampleservice-73e3607e7980ff4b57cb2c455cd0af89", :service-name "example-service", :instance-port 10005, :method "GET", :cid "262aa11d38d4c3-4b96fc8e9c07101", :total-latency 360, :discovery-latency 2, :instance-id "262a9dbad2d076-16de7b48e62c412c", :instance-latency 3, :host "example.com", :status 502, :instance-proto "http", :principal "sradack", :timestamp "2018-02-22T09:04:06.693-06:00", :scheme :http, :instance-host "127.0.0.5", :service-version "2", :metric-group "example-metric-group"}`

`{:termination-state :success, :path "/hello", :service-id "waiter-service-exampleservice-b933688e1e6069bbfddb4a4cca03036b", :service-name "example-service", :backend-latency 7, :instance-port 10007, :method "GET", :cid "262abdb4df5cd6-246211cc3f4119", :total-latency 3019, :bytes-streamed 11, :discovery-latency 10, :instance-id "262abdb67cfb94-53d3497ab0b9bce8", :overhead-latency 3012, :instance-latency 3011, :host "example.com", :status 200, :instance-proto "http", :principal "sradack", :timestamp "2018-02-22T09:06:09.497-06:00", :scheme :http, :instance-host "127.0.0.5", :service-version "3", :metric-group "example-metric-group"}`

### CODE REVIEW

File: waiter/src/waiter/request_log.clj [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+      instance-port (assoc :instance-port instance-port)
+      instance-proto (assoc :instance-proto instance-proto)
+      metric-group (assoc :metric-group metric-group)
+      query-string (assoc :query-string query-string)
+      service-id (assoc :service-id service-id)
+      service-name (assoc :service-name service-name)
+      service-version (assoc :service-version service-version)
+      received (assoc :timestamp (-> received
+                                     (coerce/from-long)
+                                     (utils/date-to-str)))
```
Comment (dposada) [inline waiter/src/waiter/request_log.clj]: Nitpick: you should be able to leave off the parens around `coerce/from-long` and `utils/date-to-str`

Comment (dposada) [inline waiter/src/waiter/request_log.clj]: Nitpick: might be good collapsed to one line

File: waiter/src/waiter/process_request.clj [hunk 2]
```diff
@@ -376,7 +378,7 @@
                           (histograms/update! (metrics/service-histogram service-id "response-size") bytes-streamed)
                           (log/error "Error occurred after streaming" bytes-streamed "bytes in response.")
                           ; Handle lower down
-                          (throw e)))]
+                          (throw (ex-info "Error occurred during streaming" {:bytes-streamed bytes-streamed} e))))]
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Is this related to the request log?

File: waiter/src/waiter/process_request.clj [hunk 3]
```diff
@@ -385,7 +387,13 @@
                               (statsd/inc! metric-group "response_bytes" unreported-bytes)
                               bytes-streamed')
                             bytes-reported-to-statsd))]
-                    (when more-bytes-possibly-available?
+                    (if-not more-bytes-possibly-available?
+                      (let [request (-> request
+                                        (utils/mark-request-time :closed))]
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: I'm not convinced the `->` is worth it here

File: waiter/src/waiter/process_request.clj [hunk 4]
```diff
@@ -385,7 +387,13 @@
...
+                        (rlog/log (assoc (rlog/request->context request)
+                                         :bytes-streamed bytes-streamed
+                                         :status (:status response)
+                                         :termination-state :success)))
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Why not have `rlog/log` take in the `request` directly and call `rlog/request->context` instead of making the caller do it?

File: waiter/src/waiter/process_request.clj [hunk 5]
```diff
@@ -385,7 +387,13 @@
                               (statsd/inc! metric-group "response_bytes" unreported-bytes)
                               bytes-streamed')
                             bytes-reported-to-statsd))]
-                    (when more-bytes-possibly-available?
+                    (if-not more-bytes-possibly-available?
+                      (let [request (-> request
+                                        (utils/mark-request-time :closed))]
+                        (rlog/log (assoc (rlog/request->context request)
+                                         :bytes-streamed bytes-streamed
+                                         :status (:status response)
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Can we destructure the `status` instead?

File: waiter/src/waiter/process_request.clj [hunk 6]
```diff
@@ -397,7 +405,14 @@
               (when-let [output-stream @output-stream-atom]
                 (log/info "invoking poison pill directly on output stream")
                 (poison-pill-function output-stream))))
-          (log/error e "Exception occurred while streaming response for" service-id))
+          (log/error e "Exception occurred while streaming response for" service-id)
+          (let [{:keys [bytes-streamed]} (ex-data e)
+                request (-> request
+                            (utils/mark-request-time :closed))]
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Maybe without `->`

File: waiter/src/waiter/process_request.clj [hunk 7]
```diff
@@ -397,7 +405,14 @@
...
+            (rlog/log (assoc (rlog/request->context request)
+                             :bytes-streamed bytes-streamed
+                             :status (:status response)
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Destructure?

File: waiter/src/waiter/process_request.clj [hunk 8]
```diff
@@ -520,6 +535,12 @@
             (confirm-live-connection-without-abort)
             (let [{:keys [service-id service-description] :as descriptor} (request->descriptor-fn request)
                   {:strs [metric-group]} service-description
+                  request (-> request
+                              (assoc :metric-group metric-group
+                                     :service-id service-id
+                                     :service-name (get service-description "name")
+                                     :service-version (get service-description "version"))
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Can we destructure these guys?

File: waiter/src/waiter/process_request.clj [hunk 9]
```diff
@@ -562,13 +583,21 @@
...
                               instance (<? (prepare-instance instance-rpc-chan service-id reason-map start-new-service-fn request-state-chan
                                                              unable-to-create-instance queue-timeout-ms reservation-status-promise metric-group
-                                                             add-debug-header-into-response!))]
+                                                             add-debug-header-into-response!))
+                              request (-> request
+                                          (utils/mark-request-time :instance-reserved)
+                                          (assoc :instance-id (subs (:id instance) (inc (count service-id)))
+                                                 :instance-host (:host instance)
+                                                 :instance-port (:port instance)
+                                                 :instance-proto (:protocol instance)))]
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Can we destructure these guys?

File: waiter/src/waiter/process_request.clj [hunk 10]
```diff
@@ -590,17 +619,37 @@
...
+                                                 :status (:status response)
+                                                 :termination-state :backend-error))
+                                response))))))
                     (catch Exception e
-                      ; Handle error lower down
-                      (throw (ex-info "Rethrowing exception from process" {:descriptor descriptor :source :pr/process} e)))))))
+                      (let [{:keys [descriptor]} (ex-data e)
+                            response (-> (process-exception-fn track-process-error-metrics request descriptor e)
+                                         (update :headers (fn [headers]
+                                                            (merge @response-headers headers))))
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: What's the effect of using `process-exception-fn` instead of `throw`ing?

File: waiter/src/waiter/process_request.clj [hunk 11]
```diff
@@ -590,17 +619,37 @@
...
+              (let [{:keys [descriptor]} (ex-data e)
+                    response (-> (process-exception-fn track-process-error-metrics request descriptor e)
+                                 (update :headers (fn [headers]
+                                                    (merge @response-headers headers))))
+                    request (-> request
+                                (utils/mark-request-time :closed))]
+                (rlog/log (assoc (rlog/request->context request)
+                                 :status (:status response)
+                                 :termination-state :service-discovery-error))
+                response))))))))
```
Comment (dposada) [inline waiter/src/waiter/process_request.clj]: Seems like the same code appears 3 times, with maybe the only variant being the `:termination-state` -- should we extract a function for this?

File: waiter/src/waiter/request_log.clj [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+;;
+;;       The copyright notice above does not evidence any
+;;       actual or intended publication of such source code.
+;;
+(ns waiter.request-log
+  (:require [clj-time.coerce :as coerce]
+            [clojure.string :as str]
+            [waiter.utils :as utils]))
+
+(def ^:dynamic *request-log-file* "log/request.log")
```
Comment (dposada) [inline waiter/src/waiter/request_log.clj]: Can we make this configurable?

File: waiter/src/waiter/request_log.clj [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+
+(defn log
+  "Log a request context."
+  [context]
+  (spit *request-log-file* (str context \newline) :append true))
```
Comment (dposada) [inline waiter/src/waiter/request_log.clj]: Should we use the logging library to do this?

File: waiter/src/waiter/utils.clj [hunk 14]
```diff
@@ -531,3 +531,8 @@
                (str (digest/digest "MD5" (str/join "" (persistent! acc))))))]
     (log/debug "got ID" id "for" sorted-parameters)
     id))
+
+(defn mark-request-time
+  "Mark the completion of a timing event with the current time in the request timing map."
+  [request k]
+  (assoc-in request [:timing k] (System/currentTimeMillis)))
```
Comment (shamsimam) [inline waiter/src/waiter/utils.clj]: Should we use `t/now` instead?

File: waiter/src/waiter/core.clj [hunk 15]
```diff
@@ -120,35 +120,37 @@
...
 
 (defn websocket-handler-factory
   "Creates the handler for processing websocket requests.
    Websockets are currently used for inter-router metrics syncing."
   [{:keys [default-websocket-handler-fn router-metrics-handler-fn]}]
   (fn websocket-handler [{:keys [uri] :as request}]
-    (case uri
-      "/waiter-router-metrics" (router-metrics-handler-fn request)
-      (default-websocket-handler-fn request))))
+    (let [request (utils/mark-request-time request :received)]
```
Comment (shamsimam) [inline waiter/src/waiter/core.clj]: We now need to be able to exclude data introduced via these calls whenever we log  the request.

Comment (dposada) [reply waiter/src/waiter/core.clj]: Do we need to exclude the data?

File: waiter/src/waiter/process_request.clj (hunk 2 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Yes.  The exception handler wants `bytes-streamed` so it can report bytes streamed so far.

File: waiter/src/waiter/process_request.clj (hunk 8 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: `name` is a function.  It's actually used inside the current `let` block which is why avoided the destructuring.

File: waiter/src/waiter/process_request.clj (hunk 9 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: We can.  It feels awkward to me, but `id` `name` `port` and `protocol` are all available.

Comment (dposada) [reply waiter/src/waiter/process_request.clj]: Why awkward?

File: waiter/src/waiter/core.clj (hunk 15 above)
Comment (shamsimam) [reply waiter/src/waiter/core.clj]: I think so, we may not need to view everything in here in our standard logs if we ever log the request.

**Comment by shamsimam** [comment]
We need to trigger a write of the log at https://github.com/twosigma/waiter/pull/208/files#diff-d77fcc9f68a157c1df598bfad7a0ccf4R552

**Comment by shamsimam** [comment]
The backend response time is available at https://github.com/twosigma/waiter/blob/17e758481ce01be829847b4c2a75f509ccb13fde/waiter/src/waiter/process_request.clj#L604 , maybe it belongs in the request log?

File: waiter/src/waiter/request_log.clj (hunk 13 above)
Comment (sradack) [reply waiter/src/waiter/request_log.clj]: Yes.

File: waiter/src/waiter/request_log.clj (hunk 1 above)
Comment (sradack) [reply waiter/src/waiter/request_log.clj]: Done

Comment (sradack) [reply waiter/src/waiter/request_log.clj]: Done

File: waiter/src/waiter/process_request.clj (hunk 3 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Done

File: waiter/src/waiter/process_request.clj (hunk 4 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Done

File: waiter/src/waiter/process_request.clj (hunk 5 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Done

File: waiter/src/waiter/process_request.clj (hunk 6 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Done

File: waiter/src/waiter/process_request.clj (hunk 7 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Done

File: waiter/src/waiter/process_request.clj (hunk 9 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: The instance fields are now in scope for a larger function.  I did it anyway, and it feels less awkward.

File: waiter/src/waiter/process_request.clj (hunk 10 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Lots has changed here.  Not sure if the question still applies.

File: waiter/src/waiter/process_request.clj (hunk 11 above)
Comment (sradack) [reply waiter/src/waiter/process_request.clj]: Done

File: waiter/src/waiter/utils.clj (hunk 14 above)
Comment (sradack) [reply waiter/src/waiter/utils.clj]: Done

File: waiter/src/waiter/core.clj (hunk 15 above)
Comment (sradack) [reply waiter/src/waiter/core.clj]: OK we can do an exclude pass when we're happy with everything that's in the log.

File: waiter/src/waiter/metrics.clj [hunk 16]
```diff
@@ -449,6 +449,16 @@
...
 
+(defmacro with-timer
+  "Times the operation specified by body using the provided
+  timer and calls handle-elapsed-nanos-fn with the epased
+  time in nanoseconds. Returns [value of body, nanos]."
+  [timer & body]
+  `(let [^Timer$Context start# (metrics.timers/start ~timer)
+         out# (do ~@body)
+         elapsed# (metrics.timers/stop start#)]
+     [out# elapsed#]))
```
Comment (shamsimam) [inline waiter/src/waiter/metrics.clj]: Why not return in a map?

File: waiter/src/waiter/request_log.clj [hunk 17]
```diff
@@ -0,0 +1,57 @@
...
+      received (assoc :timestamp (utils/date-to-str received))
+      (and sent-to-backend closed) (assoc :backend-latency (t/in-millis (t/interval sent-to-backend closed)))
+      (and received service-discovered) (assoc :discovery-latency (t/in-millis (t/interval received service-discovered)))
+      (and received instance-reserved) (assoc :instance-latency (t/in-millis (t/interval received instance-reserved)))
+      (and received closed) (assoc :total-latency (t/in-millis (t/interval received closed))))))
+
+(defn log-request
+  "Logs a request and any additional context."
+  [request & {:as additional-context}]
+  (log (merge (request->context request) additional-context)))
```
Comment (shamsimam) [inline waiter/src/waiter/request_log.clj]: Shoul;d we merge in `additional-context` into the request before generating  the context? I think we shouldn't assume that there is no data in `additional-context` (maybe rename the value) that will be used to build the context.

File: waiter/src/waiter/process_request.clj [hunk 18]
```diff
@@ -611,8 +646,9 @@
                                time (assoc :suspended-at (utils/date-to-str time))
                                (not (str/blank? last-updated-by)) (assoc :last-updated-by last-updated-by))]
       (log/info "Service has been suspended" response-map)
-      (track-response-status-metrics service-id service-description 503)
+      (track-response-status-metrics request 503)
       (meters/mark! (metrics/service-meter service-id "response-rate" "error" "suspended"))
+      (rlog/log-request request :status 503 :termination-state :service-suspended)
```
Comment (shamsimam) [inline waiter/src/waiter/process_request.clj]: Should we assume that this is a terminal operation for processing the response and trigger `log-request`? 
`process` is in control of what response get written and is better positioned to make the `log-request` call.
